src/promql/src/extension_plan/empty_metric.rs

```diff
@@ -0,0 +1,372 @@
+            end: self.end,
+            interval: self.interval,
+            schema: Arc::new(self.schema.as_ref().into()),
+            metric: ExecutionPlanMetricsSet::new(),
+        })
+    }
+}
+
+impl UserDefinedLogicalNode for EmptyMetric {
+    fn as_any(&self) -> &dyn Any {
+        self as _
+    }
+
+    fn inputs(&self) -> Vec<&LogicalPlan> {
+        vec![]
+    }
+
+    fn schema(&self) -> &DFSchemaRef {
+        &self.schema
+    }
+
+    fn expressions(&self) -> Vec<Expr> {
+        vec![]
+    }
+
+    fn fmt_for_explain(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "EmptyMetric: range=[{}..{}], interval=[{}]",
+            self.start, self.end, self.interval,
+        )
+    }
+
+    fn from_template(
+        &self,
+        _exprs: &[datafusion::prelude::Expr],
+        _inputs: &[LogicalPlan],
+    ) -> Arc<dyn UserDefinedLogicalNode> {
+        Arc::new(self.clone())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct EmptyMetricExec {
+    start: Millisecond,
+    end: Millisecond,
+    interval: Millisecond,
+    schema: SchemaRef,
+
+    metric: ExecutionPlanMetricsSet,
+}
+
+impl ExecutionPlan for EmptyMetricExec {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        None
+    }
+
+    fn maintains_input_order(&self) -> bool {
+        true
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(self.as_ref().clone()))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        _context: Arc<TaskContext>,
+    ) -> DataFusionResult<SendableRecordBatchStream> {
+        let baseline_metric = BaselineMetrics::new(&self.metric, partition);
+        Ok(Box::pin(EmptyMetricStream {
+            start: self.start,
+            end: self.end,
+            interval: self.interval,
+            is_first_poll: true,
+            schema: self.schema.clone(),
+            metric: baseline_metric,
+        }))
+    }
+
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default => write!(
+                f,
+                "EmptyMetric: range=[{}..{}], interval=[{}]",
+                self.start, self.end, self.interval,
+            ),
+        }
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metric.clone_inner())
+    }
+
+    fn statistics(&self) -> Statistics {
+        let estimated_row_num = (self.end - self.start) as f64 / self.interval as f64;
```

Comment:
Is it necessary to limit the maximum row number?

Reply:
This is only for statistics. 

The problem that too many rows are calculated inside one batch should be fixed in someplace else, but I haven't figured out how to do it properly 🤧